msgs: use effect-based diagnostic emission/handling

zerbina opened the pull request:
## Summary

* report errors/warnings/hints in the order they're created, fixing the
  out-of-order reporting
* make emission and handling of diagnostics effect-based

## Details

The current deferred approach to diagnostics, where most sem errors are
first placed into AST and then later emitted/reported in bulk has the
major problem of requiring the errors to first "bubble up", which
requires unnecessary work in some cases, also leading to out-of-order
reporting when mixed with what was previously considered
legacy reporting.

Emitting error diagnostics immediately allows handling them right away
(e.g., by terminating the program). Emission is modeled as an effect,
which is approximated via a callback.

### High-Level Description

* `emit` dispatches to the current handler, which is stored in
  `ConfigRef` for now, as that's already available everywhere
  diagnostics are emitted
* the high-level routines previously doing reporting now emit the
  diagnostic instead
* `newError` is treated as committing to the diagnostic, emitting
  it directly
* places that need to override diagnostic handling now override the new
  *diagnostic handler* not the *report hook*
* the default diagnostic handler sends the diagnostic to the report
  hook, same as before
* since `Report` is a superset of `PAstDiag`, `emit` uses `Report` to
  represent diagnostics

### Concrete Changes

* change internal error emission (`internalError` and `internalAssert`)
  to report the diagnostic directly and then quit
* change `localReport` to emit diagnostics
* change `globalReport` to always raise an exception
* remove the concept of diagnostic/report writability (it's obsolete)
* remove all after-the-fact emission of errors stored in ASTs
* remove all manual `compiles` short-circuiting logic
* remove the now-obsolete glue module `errorreporting`
* remove the now-unused/obsolete `doRaise` error action
* remove the now-obsolete `nimDebugUnreportedErrors` debug condsym
* adjust the `docgen` message handler to only use `globalReport` for
  locally-fatal errors, `localReport` otherwise

Places previously relying on `nkError` nodes for error recovery are
changed to use the diagnostic handler:
* overload resolution (diagnostic emitting during analysis are recorded
  in a separate container)
* `compiles` analysis (`tryExpr`)
* checking whether expressions are constant (`tryConstExpr`)
* call-operator-as-fallback handling; the relevant code in
  `semIndirectCall` is refactored
* `parseCode` macro API implementation (`vmdeps.parseCode`)

### Tests

Error messages and order changed in a few cases, and the expected
output of relevant tests are updated accordingly.

zerbina committed:
remove the `rsemCustomGlobalError` report

It's unused.

zerbina committed:
introduce a diagnostic handler

* add the `msgs.emit` procedure, which invokes the diagnostic handler
* `localReport` doesn't call `handleReport` anymore, but instead
  emits the diagnostic/report

zerbina committed:
add and use a default diagnostic handler

Add a default diagnostic handler that just dispatches to `handleReport`.

zerbina committed:
msgs: remove unused routines

zerbina committed:
msgs: remove `doRaise`

It's not used anymore.

zerbina committed:
errorhandling: emit diagnostics immediately

Diagnostics are now emitted immediately on constructing an `nkError`
node with them.

zerbina committed:
msgs: remove `eh` parameter from `handleReport`

It's unused.

zerbina committed:
msgs: don't use `handleReport` for internal errors

Report the diagnostic and then quit immediately, giving full control
over what happens to `doInternalXXX`.

zerbina committed:
remove the `nimDebugUnreportedErrors` condsym

Since errors are now reported right at their creation, the debug
feature is obsolete.

zerbina committed:
sem: remove `localReport`ing of `nkError` nodes

Errors are now emitted upon constructing the node, meaning that
manually emitting them is no longer necessary. The `errorreporting`
module is not used anymore and thus removed.

zerbina committed:
sigmatch: implement diagnostics recording

`untyped` parmaters allow for error recovery (internally), meaning
that errors (and other diagnostics) emitted during argument analysis
cannot be optimistically passed on to the outer handler, as a later-
checked overload might have an `untyped` formal parameter.

Therefore, diagnostics for arguments during overload resolution have
to be recorded and only be played back when their corresponding
typed argument is used by the winning overload.

zerbina committed:
sem: make `tryConstExpr` work again

Instead of relying on the global error counter, `tryConstExpr` now
overrides the diagnostics handler with one that aborts the delimited
continuation whenever an error happens, while recording all other
diagnostics (for later emission).

zerbina committed:
sem: use diagnostic effect handler for `tryExpr`

zerbina committed:
sem: remove output-based error emissions special handling

Some error emission explicitly used `globalReport` when in a `compiles`
context in order to improve latency. With the new diagnostic handling,
this is no longer necessary.

zerbina committed:
vmdeps: wrap parsing in a diagnostic handler

zerbina committed:
semexprs: use effect handler for call operator fallback

The call operator is only tried when fall-back overload resolution
fails. The code is restructured such that it's both simpler and works
with diagnostic effects.

zerbina committed:
semcall: fix "undeclared field" error not being reported

The error was erroneously captured and thus effectively swallowed.

zerbina committed:
pragmas: fix error correction

Error were eagerly created for invalid pragmas and then discarded
later, something which doesn't work anymore. Instead, the relevant
pragma processing now invokes an "invalid pragma" handler, which either
emits an error, or simply returns "empty" (to ignore the pragma).

zerbina committed:
msgs: don't overwrite existing message context

Message context needs a bigger rethink in general, but for now this
suffices.

zerbina committed:
sigmatch: fix default handling diagnostics

Diagnostics reported during parameter default handling are specific to
the current candidate and must also be registered with it.

This change introduce a "general recording mode", preparing for
recording other candidate-specific diagnostics.

zerbina committed:
sigmatch: remove early-out due to error for `matchesAux`

Overload resolution can still make progress even when `paramTypesMatch`
emits an error (possible for, e.g., late lambda expression resolution).
`noMatchDueToError` and its usages are removed.

zerbina committed:
semcall: fix error recovery via dot calls

Dot call resolution "shifts" the arguments to the right, but the slots
didn't relect this. The slot list is now also shifted.

zerbina committed:
sigmatch: fix diagnostics not being emitted

When there are no matches (e.g., all candidates have the wrong arity),
diagnostics emitted for arguments were effectively swallowed.

zerbina committed:
docgen: update RST event handling

`globalReport` always raises an exception now, meaning that it cannot
be used for emitting warnings and hints anymore. Hints and warnings use
`localReport` now.

zerbina committed:
tests: update message order in tests

zerbina committed:
tests: update expected test outputs

The changes to emission leads to some error messages being slightly
different.

zerbina committed:
tests: mark an IC test as a known error

For reasons unknown (meh), the compiler crashes on reporting a
"not found" error. Given that the previous compiler didn't reach
creation of this error (due to error-output-based early termination),
it doesn't seem unlikely that the problem is pre-existing.

zerbina commented:
The change is not perfect. There are undoubtedly some bugs it introduces, or at least error messages that become worse.

Having to record diagnostics makes sigmatch a bit more complicated and reasoning about the matching somewhat harder, and I did experiment with introducing an `nkDiags` node (which wrapped a tree with the diagnostics recorded for it) to improve on this.

The `nkDiags` approach made recording and associating diagnostics with overload candidates a bit easier, but introducing a new meta node kind that has to be handled in a lot of case statement seemed like it was too much. In addition, installing an effect handler per-operand also has a significant performance overhead.

Despite the problems, I do think that this PR is an improvement.

zerbina committed:
Merge branch 'devel' into sem-use-report-handler

zerbina committed:
Merge branch 'devel' into sem-use-report-handler

saem commented on compiler/front/msgs.nim:
```diff
@@ -644,94 +635,78 @@ proc handleReport*(
+
+proc report(config: ConfigRef, report: sink Report,
+            at: InstantiationInfo) {.inline.} =
+  ## Sends `report` to the report hook and discards the result.
```
```suggestion
  ## Sends `report` to the report hook and discards the result.
  ## xxx `report` is a legacy name and needs to change
```

I want to ensure we don't codify the vocabulary, I still firmly believe the current "report" plague needs to be excised from the compiler.

zerbina replied:
What's the terminology you want to replace "report" with (in this specific case, not generally)? I'm currently using "report" to mean "render/show/display". Essentially, `report` is called when all internal handling (e.g., recording, discarding, suppression, etc.) for a diagnostic is done and it now needs to be sent to the display engine in a fire-and-forget manner. Maybe `display` or `show` would fit better here?

To be be clear, I have no problem with adding the comment.

saem replied:
I was thinking we'd call it `log`, although maybe I'll have less of a strong reaction to `report` once the actual legacy code is removed.

saem commented on compiler/tools/docgen.nim:
```diff
@@ -253,27 +253,35 @@ template declareClosures =
+      of mwRstStyle:                rbackRstRstStyle
+    if kind in rbackErrorKinds:
+      # TODO: errors/events that are locally fatal have to handle aborting
+      #       where the event is emitted, not here via `globalReport`
```
Testing my understanding: this means that various parts of docgen should be updated to use locally fatal error reporting, rather than relying on this catchall, correct?

zerbina replied:
Yea, that's what I meant. Either there needs to be a dedicated "fatal error" noreturn handler, or the respective callsites need to raise an exception themselves.

The handler implementation here should just send the diagnostic to the reporting/display management, without also doing control flow. If it does do control flow, that's totally fine, but it should not be *required* to.

saem reviewed (approved):
That's a huge readability improvement, it gets rid of a lot of knots in the compiler (compiles, sigmatch, error kind handling, writability, etc).

saem commented:
/merge

saem commented:
I've taken a peek at the failure (leak detection), but I can't figure out why it's happening

log: https://github.com/nim-works/nimskull/actions/runs/17535039994/job/49796993395

zerbina committed:
work around a compiler bug

Add a setter for the diagnostic handler and replace all manual
assignments with it. The lambda expressions being passed to a sink
parameter ensures the hooks for the closure type are properly
synthesized (something that currently doesn't happen on direct
assignments).

zerbina committed:
semcall: optimize handler setup

Move pushing the handler into a separate procedure, to get around
copying the env reference. Instead, the diagnostic context ref is
copied, which should be less problematic, as there are less heap cells
reachable from it.

zerbina commented:
> I've taken a peek at the failure (leak detection), but I can't figure out why it's happening

There's an issue with `sempass2` that leads to the lambda expressions' type not having its lifetime hooks synthesized properly. I've worked around the issue by using a dedicated setter routine, which has shown to fix the leak when I tested it locally (though I might have missed some occurrences).

saem commented:
/merge
